src/core/trace_core.mli

```diff
@@ -11,43 +12,60 @@ val enabled : unit -> bool
     This is fast, so that the traced program can check it before creating
     any span or message *)
 
-val enter_span :
+val with_span :
   ?__FUNCTION__:string ->
   __FILE__:string ->
   __LINE__:int ->
   ?data:(unit -> (string * user_data) list) ->
   string ->
-  span
-(** Enter a span. A span is a delimited period of time with
-    a start instant ("enter") and stop instant ("exit"), associated
-    with some metadata about the code entering/exiting some piece of code.
-
-    In particular the entrypoint comes with the location of the code
-    (you can use [__FILE__] and [__LINE__] directly in OCaml),
-    a mandatory name, and some optional metadata in a JSON-like representation.
-*)
+  (span -> 'a) ->
+  'a
+(** [with_span ~__FILE__ ~__LINE__ name f] enters a new span [sp],
+    and calls [f sp].
+    [sp] might be a dummy span if no collector is installed.
+    When [f sp] returns or raises, the span [sp] is exited.
 
-val exit_span : span -> unit
-(** Exit the span.
+    This is the recommended way to instrument most code.
 
-    This should be called exactly once per span (even in case of exception).
-    Once this is called, a timestamp might be recorded,
-  and the span is considered finished and can't be used anymore *)
+    {b NOTE} an important restriction is that this is only supposed to
+    work for synchronous, direct style code. Monadic concurrency, Effect-based
+    fibers, etc. might not play well with this style of spans on some
+    or all backends. If you use cooperative concurrency,
+    see {!enter_manual_span}.
+*)
 
-val with_span :
+val enter_manual_sub_span :
+  parent:explicit_span ->
   ?__FUNCTION__:string ->
   __FILE__:string ->
   __LINE__:int ->
   ?data:(unit -> (string * user_data) list) ->
   string ->
-  (span -> 'a) ->
-  'a
-(** [with_span ~__FILE__ ~__LINE__ name f] enters a new span [sp],
-    and calls [f sp].
-    [sp] might be a dummy span if no collector is installed.
-    When [f sp] returns or raises, the span [sp] is exited.
+  explicit_span
+(** Like {!with_span} but the caller is responsible for
+    providing the [parent] context, and carry the resulting
```

Comment:
ditto — i'd make the 'by exposing it in your library API' part explicit. give library-authors explicit directions, so they don't have to think:

```suggestion
    obtaining the [parent] span from their {e own} caller, and carry the resulting
```